File: apps/desktop/src/services/backend.ts
/**
 * Tauri 后端托管 IPC 服务。
 *
 * 统一封装桌面端对 Rust 本地后端托管命令的调用，
 * 组件和 Hook 不直接触碰 `invoke()`。
 *
 * @module services/backend
 */

import type { BackendLogsTailResponse, BackendStatusResponse } from "@shared/backend";
import { ServiceError } from "./types";
import { logError } from "../lib/logger";

/**
 * 调用 Tauri IPC 命令并统一映射错误。
 *
 * @param command - Rust 侧注册的命令名。
 * @param args - 可选命令参数。
 * @returns 命令的结构化返回值。
 * @throws {ServiceError} 当 IPC 调用失败时抛出。
 */
async function invokeCommand<T>(command: string, args?: Record<string, unknown>): Promise<T> {
  try {
    const { invoke } = await import("@tauri-apps/api/core");
    return await invoke<T>(command, args);
  } catch (error) {
    logError(`Tauri IPC 调用失败: ${command}`, error, {
      module: "backendService",
      command,
    });
    throw new ServiceError(`Tauri IPC 调用失败: ${command}`, {
      cause: error,
    });
  }
}

/**
 * 启动桌面端托管的本地 Python 后端。
 *
 * @returns 启动后的结构化后端状态。
 * @throws {ServiceError} 当 IPC 通道调用失败时抛出。
 */
export async function startBackend(): Promise<BackendStatusResponse> {
  return invokeCommand<BackendStatusResponse>("backend_start");
}

/**
 * 停止桌面端托管的本地 Python 后端。
 *
 * @returns 停止后的结构化后端状态。
 * @throws {ServiceError} 当 IPC 通道调用失败时抛出。
 */
export async function stopBackend(): Promise<BackendStatusResponse> {
  return invokeCommand<BackendStatusResponse>("backend_stop");
}

/**
 * 重启桌面端托管的本地 Python 后端。
 *
 * @returns 重启后的结构化后端状态。
 * @throws {ServiceError} 当 IPC 通道调用失败时抛出。
 */
export async function restartBackend(): Promise<BackendStatusResponse> {
  return invokeCommand<BackendStatusResponse>("backend_restart");
}

/**
 * 查询桌面端托管的本地 Python 后端状态。
 *
 * @returns 当前结构化后端状态。
 * @throws {ServiceError} 当 IPC 通道调用失败时抛出。
 */
export async function getBackendStatus(): Promise<BackendStatusResponse> {
  return invokeCommand<BackendStatusResponse>("backend_status");
}

/**
 * 读取本地后端相关日志尾部内容。
 *
 * @param maxLines - 每个日志文件最多返回的行数。
 * @returns 多个日志文件的尾部内容。
 * @throws {ServiceError} 当 IPC 通道调用失败时抛出。
 */
export async function tailBackendLogs(maxLines = 80): Promise<BackendLogsTailResponse> {
  return invokeCommand<BackendLogsTailResponse>("backend_logs_tail", { maxLines });
}

File: apps/desktop/src-tauri/src/commands/backend.rs
//! 本地后端托管命令。

use crate::backend::supervisor::BackendSupervisorState;
use crate::backend::types::{BackendLogsTailResponse, BackendStatusResponse};
use tauri::State;

/// 异步启动本地 Python 后端。
///
/// 参数:
///     app: Tauri 应用句柄。
///     supervisor: 后端托管状态。
///
/// 返回:
///     启动后的结构化后端状态。
///
/// 异常:
///     当后端启动路径无法解析等不可恢复错误发生时返回错误字符串。
///
/// 副作用:
///     可能拉起本地 Python 子进程并异步轮询健康检查；等待期间让出执行权，不阻塞主线程。
#[tauri::command]
pub async fn backend_start(
    app: tauri::AppHandle,
    supervisor: State<'_, BackendSupervisorState>,
) -> Result<BackendStatusResponse, String> {
    supervisor.start(&app).await
}

/// 异步停止当前由桌面端托管的本地 Python 后端。
///
/// 参数:
///     app: Tauri 应用句柄。
///     supervisor: 后端托管状态。
///
/// 返回:
///     停止后的结构化后端状态。
///
/// 异常:
///     当后端停止路径无法解析等不可恢复错误发生时返回错误字符串。
///
/// 副作用:
///     可能终止已托管的 Python 子进程，并访问 `/health` 确认停止结果。
#[tauri::command]
pub async fn backend_stop(
    app: tauri::AppHandle,
    supervisor: State<'_, BackendSupervisorState>,
) -> Result<BackendStatusResponse, String> {
    supervisor.stop(&app).await
}

/// 异步重启当前由桌面端托管的本地 Python 后端。
///
/// 参数:
///     app: Tauri 应用句柄。
///     supervisor: 后端托管状态。
///
/// 返回:
///     重启后的结构化后端状态。
///
/// 异常:
///     当后端重启路径无法解析等不可恢复错误发生时返回错误字符串。
///
/// 副作用:
///     可能终止旧的 Python 子进程并异步拉起新进程、轮询健康检查。
#[tauri::command]
pub async fn backend_restart(
    app: tauri::AppHandle,
    supervisor: State<'_, BackendSupervisorState>,
) -> Result<BackendStatusResponse, String> {
    supervisor.restart(&app).await
}

/// 异步查询本地后端当前状态。
///
/// 参数:
///     app: Tauri 应用句柄。
///     supervisor: 后端托管状态。
///
/// 返回:
///     当前结构化后端状态。
///
/// 异常:
///     当后端状态路径无法解析等不可恢复错误发生时返回错误字符串。
///
/// 副作用:
///     会探测现有进程并访问 `/health` 端点。
#[tauri::command]
pub async fn backend_status(
    app: tauri::AppHandle,
    supervisor: State<'_, BackendSupervisorState>,
) -> Result<BackendStatusResponse, String> {
    supervisor.status(&app).await
}

/// 异步读取后端相关日志的尾部内容。
///
/// 参数:
///     app: Tauri 应用句柄。
///     supervisor: 后端托管状态。
///     max_lines: 每个日志文件最多返回的尾部行数。
///
/// 返回:
///     每个日志文件的尾部文本。
///
/// 异常:
///     当后端日志路径无法解析或日志文件读取失败时返回错误字符串。
///
/// 副作用:
///     读取本地日志文件；异步持锁期间让出执行权，不阻塞主线程。
#[tauri::command]
pub async fn backend_logs_tail(
    app: tauri::AppHandle,
    supervisor: State<'_, BackendSupervisorState>,
    max_lines: Option<usize>,
) -> Result<BackendLogsTailResponse, String> {
    supervisor.logs_tail(&app, max_lines.unwrap_or(80)).await
}
